galley/pkg/config/analysis/analyzers/auth/serviceroleservices.go

```diff
@@ -0,0 +1,95 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package auth
+
+import (
+	"istio.io/api/rbac/v1alpha1"
+	"istio.io/istio/galley/pkg/config/analysis"
+	"istio.io/istio/galley/pkg/config/analysis/analyzers/util"
+	"istio.io/istio/galley/pkg/config/analysis/msg"
+	"istio.io/istio/galley/pkg/config/meta/metadata"
+	"istio.io/istio/galley/pkg/config/meta/schema/collection"
+	"istio.io/istio/galley/pkg/config/resource"
+)
+
+// ServiceRoleServicesAnalyzer checks the validity of services referred in a Service Role
+type ServiceRoleServicesAnalyzer struct{}
+
+var _ analysis.Analyzer = &ServiceRoleServicesAnalyzer{}
+
+// Metadata implements Analyzer
+func (s *ServiceRoleServicesAnalyzer) Metadata() analysis.Metadata {
+	return analysis.Metadata{
+		Name: "auth.ServiceRoleServicesAnalyzer",
+		Inputs: collection.Names{
+			metadata.IstioRbacV1Alpha1Serviceroles,
+			metadata.K8SCoreV1Services,
+		},
+	}
+}
+
+// Analyze implements Analyzer
+func (s *ServiceRoleServicesAnalyzer) Analyze(ctx analysis.Context) {
+	ctx.ForEach(metadata.IstioRbacV1Alpha1Serviceroles, func(r *resource.Entry) bool {
+		s.analyzeServiceRoleServices(r, ctx)
+		return true
+	})
+}
+
+// analyzeRoleBinding apply analysis for the service field of the given ServiceRole
+func (s *ServiceRoleServicesAnalyzer) analyzeServiceRoleServices(r *resource.Entry, ctx analysis.Context) {
+	sr := r.Item.(*v1alpha1.ServiceRole)
+	ns, _ := r.Metadata.Name.InterpretAsNamespaceAndName()
+
+	for _, rs := range sr.Rules {
+		for _, svc := range rs.Services {
+			report := false
+			rn := util.GetResourceNameFromHost(ns, svc)
+			_, ds := rn.InterpretAsNamespaceAndName()
+
+			// If service is either * or *.ns.cluster
```

Comment:
As I understand [the docs](https://istio.io/docs/reference/config/authorization/istio.rbac.v1alpha1/#AccessRule):

* `*` applies to all services in the *rule's* namespace
* `.ns.cluster.local` applies to all services in the `ns` namespace
* `*.ns.cluster.local` is invalid. (The only mention in the docs of `*` is standalone, you don't need to mark that you're doing suffix matching)
So I think the logic here needs some tweaking. I'm certainly not an expert in how the service roles are attached to services, but that's how I read the docs.

Reply:
Working on that now :)

Reply:
I've re-read the doc again and I am having doubts. 

> Exact match, prefix match, and suffix match are supported for service names

Does that mean that we should consider also cases like:

- `book` matches as prefix of `bookstore.mtv.cluster.local` and `booksmth.mtv.cluster.local`

- `tore.mtv.cluster.local` matches as suffix of `bookstore.mtv.cluster.local` and `pencilstore.mtv.cluster.local`
- `tv.cluster.local` matches as suffix of `bookstore.mtv.cluster.local` and `whatever.btv.cluster.local`

I need to dig a little big more into the sourcecode. Or do you know who I can ping to discuss about that?

Thanks

Reply:
Yeah, one way or another the docs probably need to be clearer!

I don't know who to ask, but since this is for an RBAC feature maybe it makes sense to start with the #security slack channel?

Reply:
I applied a solution that is following what is documented. It applies the prefix, suffix and equal match for those service in the same namespace. Then the cases are mentioned in my previous comment will fit in the validation.
However, on the other hand, I opened an [issue reporting a bug](https://github.com/istio/istio/issues/18327) (according the documentation). Hope that it helps to get a clearer scenario.
Anyway, the test cases in this PR covers a good amount of situations.